### DISCUSSION TITLE
Alembic doesn't recognize JSON

### DISCUSSION THREAD

**Discussion by cheog** [sqlalchemy/sqlalchemy#8987] [Usage Questions] [answered]
**Describe the bug**
Alembic doesn't recognize JSON

**Expected behavior**
I expect alembic to autodetect/support a change from a string to a json field.

**To Reproduce**
I did the following:
1. Ran `main.py` version 1 (producing test.db)
2. Did an `alembic init alembic`
3. Modified the alembic.ini and env.py
4. Modified `main.py` to version 2
5. Ran `alembic revision -m "test" --autogenerate

```py
# main.py (version 1)
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy import create_engine

class Base(DeclarativeBase):
    pass

class Chicken(Base):
    __tablename__ = "chicken"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    eggs: Mapped[list["Egg"]] = relationship(back_populates="config")

class Egg(Base):
    __tablename__ = "egg"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    chicken_id: Mapped[int] = mapped_column(ForeignKey("chicken.id"))
    chicken: Mapped["Chicken"] = relationship(back_populates="runs")

engine = create_engine("sqlite:///test.db", echo=True)
Base.metadata.create_all(engine)
```

```py
# main.py (version 2)
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.types import JSON
from sqlalchemy import create_engine

class Base(DeclarativeBase):
    pass

class Egg(Base):
    __tablename__ = "egg"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    chicken: Mapped[JSON]

engine = create_engine("sqlite:///test.db", echo=True)
Base.metadata.create_all(engine)
```

```
# alembic.ini (minus the logging config)
[alembic]
script_location = alembic
prepend_sys_path = .
version_path_separator = os
sqlalchemy.url = sqlite:///test.db
```

```py
# env.py (only the lines that were changed from the default)
from main import Base
target_metadata = Base.metadata
```

**Error**

```
Traceback (most recent call last):
  File "/Users/me/Documents/test_json/__pypackages__/3.10/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/config.py", line 590, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/config.py", line 584, in main
    self.run_cmd(cfg, options)
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/config.py", line 561, in run_cmd
    fn(
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/command.py", line 229, in revision
    script_directory.run_env()
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/script/base.py", line 569, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/me/Documents/test_json/alembic/env.py", line 5, in <module>
    from main import Base
  File "/Users/me/Documents/test_json/./main.py", line 18, in <module>
    class Egg(Base):
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/decl_api.py", line 706, in __init_subclass__
    _as_declarative(cls._sa_registry, cls, cls.__dict__)
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/decl_base.py", line 234, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/decl_base.py", line 315, in setup_mapping
    return _ClassScanMapperConfig(
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/decl_base.py", line 557, in __init__
    self._extract_mappable_attributes()
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/decl_base.py", line 1376, in _extract_mappable_attributes
    value.declarative_scan(
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/properties.py", line 680, in declarative_scan
    self._init_column_for_annotation(
  File "/Users/me/Documents/test_json/__pypackages__/3.10/lib/sqlalchemy/orm/properties.py", line 763, in _init_column_for_annotation
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type: <class 'sqlalchemy.sql.sqltypes.JSON'>
```

**Versions.**
 - OS: MacOS 13.0.1 (M1)
 - Python: 3.10.8
 - Alembic: 1.8.1
 - SQLAlchemy: 2.0.0b4
 - Database: SQLite
 - DBAPI: idk

**Additional context**
I also tried with a manual migration instead of the autogenerate and got the same error.

**Comment by CaselIT** [comment, marked as answer]
Hi,

This is not an alembic issue, is a duplicate of https://github.com/sqlalchemy/sqlalchemy/discussions/8980 and a better error message will be provided on in the next version.

tldr: Mapped requires a python type, not an sqlalchemy type

**Reply by cheog** [reply]
Thank you so much for your help!!! I do have a few follow up questions.

So the equivalent of JSON would be a pydantic dict right? I tried with a dict and it didn't work (error at the bottom). Is there a place I can look up the mapping between JSON / UUID / etc their python counterparts?

Furthermore, if I'm not using a dict and instead using something like a string then should I use a declaration something like
```py
class Egg(Base):
    __tablename__ = "egg"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    chicken: Mapped[str] = mapped_column(type_=JSON)
```
and then every time I do a query I have a `loads` step? Or is there a More Correct way to do this?

Error message for dict:  
`sqlalchemy.exc.ArgumentError: Could not locate SQLAlchemy Core type for Python type: <class 'dict'>`

**Reply by CaselIT** [reply]
So for that use case there are various solutions, documented in various point here https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-annotated-declarative-table-type-annotated-forms-for-mapped-column but since dict and list could be used also for registry it's probably best to provide the type in `mapped_column`

> Furthermore, if I'm not using a dict and instead using something like a string then should I use a declaration something like
> ...

No, the type seen by sqlalchemy in that case is still JSON since when both the type in `Mapped[]` and in `mapped_column` are provided the one in `mapped_column` wins, so sqlalchemy would still call `loads` when loading the type from the database

**Comment by CaselIT** [comment]
moving the discussion to sqlalchemy

**Comment by jgolden-efcg** [comment]
For anyone else who stumbles upon this, here's what I did:

1. Defined a type `ormJSON`:
```
class ormJSON(dict):
    pass
```

2. Added a custom mapping to the type annotation map in my base class:
```
from sqlalchemy.types import JSON
class Base(DeclarativeBase):
    registry = registry(type_annotation_map={ormJSON: JSON})
```

3. Declared mapped columns with the `ormJSON` type:
```
...
my_col: Mapped[ormJSON]
```

Seems to be working so far.

**Reply by CaselIT** [reply]
You may also try annotated instead of a subclass:
```py
from typing import Annotated
from sqlalchemy.types import JSON

ormJSON = Annotated[dict, 'anything']

class Base(DeclarativeBase):
    registry = registry(type_annotation_map={ormJSON: JSON})
```
docs at https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapping-multiple-type-configurations-to-python-types
